app/lib/params_sanitizer.rb

```diff
@@ -0,0 +1,461 @@
+    :citationsOverTime,
+    { citationsOverTime: %i[year total] },
+    :citationCount,
+    :downloadCount,
+    :partCount,
+    :partOfCount,
+    :referenceCount,
+    :versionCount,
+    :versionOfCount,
+    :viewCount,
+  ].freeze
+
+  RELATIONSHIPS_MAP = [{ client: [data: %i[type id]] }].freeze
+
+  def initialize(params = {})
+    @params = params
+  end
+
+  def cleanse
+    xml = add_xml
+
+    meta = generate_meta(xml)
+    add_schema_version(meta)
+    xml = meta["string"]
+    # if metadata for DOIs from other registration agencies are not found
+    fail ActiveRecord::RecordNotFound if meta["state"] == "not_found"
+
+    add_random_id
+
+    # replace DOI, but otherwise don't touch the XML
+    # use Array.wrap(read_attrs.first) as read_attrs may also be [[]]
+    if meta["from"] == "datacite" && Array.wrap(read_attrs.first).blank?
+      xml = replace_doi(xml, doi: @params[:doi] || meta["doi"])
+    elsif xml.present? || Array.wrap(read_attrs.first).present?
+      regenerate = true
+    end
+
+    @params[:xml] = xml if xml.present?
+
+    add_xml_attributes(meta)
+    add_metadata_version(meta)
+    add_landingpage()
+
+    @params.merge(regenerate: @params[:regenerate] || regenerate).except(
+      # ignore camelCase keys, and read-only keys
+      :confirmDoi,
+      :prefix,
+      :suffix,
+      :publicationYear,
+      :alternateIdentifiers,
+      :rightsList,
+      :relatedIdentifiers,
+      :relatedItems,
+      :fundingReferences,
+      :geoLocations,
+      :metadataVersion,
+      :schemaVersion,
+      :state,
+      :mode,
+      :isActive,
+      :landingPage,
+      :created,
+      :registered,
+      :updated,
+      :published,
+      :lastLandingPage,
+      :version,
+      :lastLandingPageStatus,
+      :lastLandingPageStatusCheck,
+      :lastLandingPageStatusResult,
+      :lastLandingPageContentType,
+      :contentUrl,
+      :viewsOverTime,
+      :downloadsOverTime,
+      :citationsOverTime,
+      :citationCount,
+      :downloadCount,
+      :partCount,
+      :partOfCount,
+      :referenceCount,
+      :versionCount,
+      :versionOfCount,
+      :viewCount,
+    )
+  end
+
+  def  read_attrs
+    [
+      @params[:creators],
+      @params[:contributors],
+      @params[:titles],
+      @params[:publisher],
+      @params[:publicationYear],
+      @params[:types],
+      @params[:descriptions],
+      @params[:container],
+      @params[:sizes],
+      @params[:formats],
+      @params[:version],
+      @params[:language],
+      @params[:dates],
+      @params[:identifiers],
+      @params[:relatedIdentifiers],
+      @params[:relatedItems],
+      @params[:fundingReferences],
+      @params[:geoLocations],
+      @params[:rightsList],
+      @params[:subjects],
+      @params[:contentUrl],
+      @params[:schemaVersion],
+    ].compact
+  end
+
+  def get_params
+    @params
+  end
+
+  def add_xml
+    # extract attributes from xml field and merge with attributes provided directly
```

Comment:
Now that things have been refactored, this comment is no longer accurate as the merging happens outside of this function, so it could maybe be reworded to reflect that or removed